Notifications android (#1814)
OWASP/owasp-mstg: Document/0x05h-Testing-Platform-Interaction.md

Next, verify whether or not you can continue to use the application without updating it. If an update prompt is given, verify if you can still use the application by canceling the prompt or otherwise circumventing it through normal application usage. This includes validating whether the backend will stop calls to vulnerable backends and/or whether the vulnerable app-version itself is blocked by the backend.
Lastly, see if you can play with the version number of a man-in-the-middled app and see how the backend responds to this (and if it is recorded at all for instance).

> ## Testing App Notifications

Reviewer: I just realised we need to give it a reference to a MSTG-ID. 

I think there are two options:

- MSTG-STORAGE-7: No sensitive data, such as passwords or pins, is exposed through the user interface.
- MSTG-STORAGE-4: No sensitive data is shared with third parties unless it is a necessary part of the architecture.

@cpholguera I would tent to reference both, as there are two attack vectors. First about exposing the data in the notification in the UI (MSTG-Storage-4) and exposing the same information to other apps that are able to read it (MSTG-Storage-7).

> ## Testing App Notifications

Reviewer 2: Agree @sushi2k ;)

```suggestion
## Testing App Notifications (MSTG-STORAGE-4 and MSTG-STORAGE-7)
```

### Overview

Android allows for applications to create [notifications](https://developer.android.com/guide/topics/ui/notifiers/notifications "Notifications Overview"), a form of message that is displayed by the Android system outside of the application's UI. This is useful for applications to display small messages to the user while they use other applications on their device.

It is important to understand that notifications should never be considered private. When a notification is handled by the Android system it is broadcasted system-wide and any application running with a [NotificationListenerService](https://developer.android.com/reference/kotlin/android/service/notification/NotificationListenerService "NotificationListenerService") can listen for these notifications to receive them in full and may handle them however it wants.

> There are many known malware samples such as [Joker](https://research.checkpoint.com/2020/new-joker-variant-hits-google-play-with-an-old-trick/ "Joker Malware"), and [Alien](https://www.threatfabric.com/blogs/alien_the_story_of_cerberus_demise.html "Alien Malware") which abuse the `NotificationListenerService` to listen for notifications on the device and then send them to attacker-controlled C2 infrastructure. Commonly this is done in order to listen for two-factor authentication (2FA) codes that appear as notifications on the device which are then sent to the attacer. A safer alterntive for the user would be to use a 2FA application that does not generate notifications.

Reviewer: ```suggestion
There are many known malware samples such as [Joker](https://research.checkpoint.com/2020/new-joker-variant-hits-google-play-with-an-old-trick/ "Joker Malware"), and [Alien](https://www.threatfabric.com/blogs/alien_the_story_of_cerberus_demise.html "Alien Malware") which abuses the `NotificationListenerService` to listen for notifications on the device and then send them to attacker-controlled C2 infrastructure. Commonly this is done in order to listen for two-factor authentication (2FA) codes that appear as notifications on the device which are then sent to the attacker. A safer alternative for the user would be to use a 2FA application that does not generate notifications.
```

Author: Opps, think my spelling check plugin wasnt working

> Furthermore there are a number of apps on the Google Play Store that provide notification logging; locally logging any notifications on the Android system. This highlights that notifications are in no way private on Android.  

Reviewer: ```suggestion
Furthermore there are a number of apps on the Google Play Store that provide notification logging, which basically logs locally any notifications on the Android system. This highlights that notifications are in no way private on Android and accessible by any other app on the device.
```

Lastly, see if you can play with the version number of a man-in-the-middled app and see how the backend responds to this (and if it is recorded at all for instance).

## Testing App Notifications (MSTG-PLATFORM-12)

### Overview

Android allows for applications to create [notifications](https://developer.android.com/guide/topics/ui/notifiers/notifications "Notifications Overview"), a form of message that is displayed by the Android system outside of the application's UI. This is useful for applications to display small messages to the user while they use other applications on their device.

It is important to understand that notifications should never be considered private. When a notification is handled by the Android system it is broadcasted system-wide and any application running with a [NotificationListenerService](https://developer.android.com/reference/kotlin/android/service/notification/NotificationListenerService "https://developer.android.com/reference/kotlin/android/service/notification/NotificationListenerService") can listen for these notifications to receive them in full. From here the Listener application can handle these notifications however it wants.

> There are many known malware examples such as Joker, and Terracotta which abuse the `NotificationListenerService` to listen for notifications on the device and then send them to attacker-controlled C2 infrastructure. Furthermore there are a number of apps on the Google Play Store that provide notification logging; locally logging any notifications on the Android system. This highlights that notifications are in no way private on Android.  

Reviewer: Can you add links to Joker or Terracotta (for example blogs that analyse them or where further information can be found)

> There are many known malware examples such as Joker, and Terracotta which abuse the `NotificationListenerService` to listen for notifications on the device and then send them to attacker-controlled C2 infrastructure. Furthermore there are a number of apps on the Google Play Store that provide notification logging; locally logging any notifications on the Android system. This highlights that notifications are in no way private on Android.  

Reviewer: ```suggestion
There are many known malware examples such as Joker, and Terracotta which abuse the `NotificationListenerService` to listen for notifications on the device and then send them to attacker-controlled Command and Control (C2) infrastructure. Furthermore there are a number of apps on the Google Play Store that provide notification logging, meaning logging any notifications coming from Apps on the Android system. This highlights that notifications are not private on Android.  
```

Consider a hypothetical application that features a two-factor authentication (2FA) function built into the application. When a user goes to sign in to the application on a web browser they are prompted for a 2FA code from their mobile device, the application creates a notification that contains this code for ease of access.

> A malicious application listening for notifications could obtain this 2FA code from the notification and send it to the attacker for use. The application should instead use the notification to prompt the user to go into the application its self to obtain the 2fa code.

Reviewer 2: What about apps like GitHub/Twitter/etc. If you login in the browser or in the app you'll get an SMS with the 2FA code, so any app will be able to see it. What should be done on that case?

> A malicious application listening for notifications could obtain this 2FA code from the notification and send it to the attacker for use. The application should instead use the notification to prompt the user to go into the application its self to obtain the 2fa code.

Author: On re-reading that first paragraph I can see why you'd be confused; made sense in my head at the time. 

Basically some applications I've tested in the past have had in built 2fa functions (Steam guard does this for example) so that when you sign into the account on a web browser a notification pops up on your phone telling you your 2fa. A malicious app can of course listen for notifications and grab this notification; thus its not just limited to 2fa on SMS.

To answer your other question 2FA over SMS is in my opinion (and the [opinion ](https://blog.sucuri.net/2020/01/why-2fa-sms-is-a-bad-idea.html )[of ](https://www.kaspersky.com/blog/2fa-practical-guide/24219/ )[many ](https://www.theverge.com/2017/9/18/16328172/sms-two-factor-authentication-hack-password-bitcoin)[others](https://www.cnet.com/how-to/do-you-use-sms-for-two-factor-authentication-heres-why-you-shouldnt/)) a weak form of 2FA for reasons both described here and inherent to SMS. 

But to focus on the Android system side and to answer what should be done in this case: Dont use sms 2fa or 2fa that generates popups.

Ultimately this is a weakness of the android system and how it handles notifications. Notifications are not private like the per app notifications of iOS and thus the best thing to be done is to avoid generating notifications.

I hope this makes sense and I'm happy to elaborate more if need be, ill also do some modifications to this section.

For this reason all notification usage should be inspected for confidential or high risk information that could be used by malicious applications.

### Static Analysis

> Notifications require the usage of the `NotificationManager` class, thus when assessing an application statically it is recommended to explore the application for any usage of this class as this will likely indicate the usage of some form of notifications.
> From here code will have to be statically analyzed in order to understand how the application is [generating the notification](https://developer.android.com/training/notify-user/build-notification#SimpleNotification "Create a Notification").

Reviewer 2: what about this shorter version?
```suggestion
When statically assessing an application, it is recommended to search for any usage of the `NotificationManager` class which might be an indication of some form of notification management. If the class is being used, the next step would be to understand how the application is [generating the notifications](https://developer.android.com/training/notify-user/build-notification#SimpleNotification "Create a Notification").
```

> These code locations can feed into the Dynamic Analysis section below, providing an idea of where in the application notificaitons may be dynamically generated.

Reviewer: ```suggestion
These code locations can be feed into the Dynamic Analysis section below, providing an idea of where in the application notifications may be dynamically generated.
```

- <https://developers.google.com/digital-asset-links/v1/getting-started>
- <https://pdfs.semanticscholar.org/0415/59c01d5235f8cf38a3c69ccee7e1f1a98067.pdf>

> ### Android App Notifcations

Reviewer 2: ```suggestion
### Android App Notifications
```

- MSTG-PLATFORM-6: "WebViews are configured to allow only the minimum set of protocol handlers required (ideally, only https is supported). Potentially dangerous handlers, such as file, tel and app-id, are disabled."
- MSTG-PLATFORM-7: "If native methods of the app are exposed to a WebView, verify that the WebView only renders JavaScript contained within the app package."
- MSTG-PLATFORM-8: "Object serialization, if any, is implemented using safe serialization APIs."
- MSTG-ARCH-9: "A mechanism for enforcing updates of the mobile app exists."

> 

Reviewer 2: ```suggestion

- MSTG-STORAGE-4: "No sensitive data is shared with third parties unless it is a necessary part of the architecture."
- MSTG-STORAGE-7: "No sensitive data, such as passwords or pins, is exposed through the user interface."

```

- MSTG-PLATFORM-7: "If native methods of the app are exposed to a WebView, verify that the WebView only renders JavaScript contained within the app package."
- MSTG-PLATFORM-8: "Object serialization, if any, is implemented using safe serialization APIs."

> - MSTG-PLATFORM-9: "The app does not allow any other application to overlay on itself."

Reviewer: ```suggestion
- MSTG-PLATFORM-9: "The app protects itself against screen overlay attacks. (Android only)"
```

> - MSTG-PLATFORM-10: "A WebView's cache, storage, and loaded resources (JavaScript, etc.) should be cleared before the WebView is destroyed."

Reviewer: We are only adding the requirements from MASVS if the test case is also added here. This one is not yet outlined here, so please remove it for now.

- MSTG-PLATFORM-11: "Verify that the app prevents usage of custom third-party keyboards whenever sensitive data is entered."

> - MSTG-PLATFORM-12: "High confidentiality information is not parsed into notifications."

Reviewer: You can find the requirements we are referencing in the MASVS here: https://github.com/OWASP/owasp-masvs/blob/master/Document/0x11-V6-Interaction_with_the_environment.md

Please remove this requirement. 

For changes to the requirements please raise an issue. For this case the following requirement might make the most sense, but will discuss with @cpholguera. What do you think?

2.7 | MSTG-STORAGE-7 | No sensitive data, such as passwords or pins, is exposed through the user interface.
-- | -- | --

Author: Will do; I wasn't sure if this needed to be modified / changed for new sections like this so I thought it'd be best I just do it anyway as worst comes to worst it can just be removed and changed back (which ill do now).
